### PR TITLE
[asio] update to v1.18.0

### PR DESCRIPTION
**Pull request by mathisloge**
**Describe the pull request**

- What does your PR fix? 
updates the library asio to version 1.18.0

- Which triplets are supported/not supported? Have you updated the CI baseline? No

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
hope so

### CODE REVIEW

File: ports/asio/portfile.cmake [hunk 1]
```diff
@@ -4,8 +4,8 @@ include(vcpkg_common_functions)
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO chriskohlhoff/asio
-    REF asio-1-12-2
-    SHA512 7c2e213ff154bb2e5776b37906d437a62206f973316c94706e6d42e3c2f0866e7d97f3e40225ab5f28bf2c4a33fa0b38a4b75421aef86ddf9f2da0811caa2d00
+    REF asio-1-18-0
+    SHA512 c79529d16a51f40c9faff8549ef7dd7074aaf2574c8c338a8885aa04b3bb261a493f57498d1bd271cfcb4083e561344e286f784c77ea20e355bf86f7a39cfb39
     HEAD_REF master
 )
```
Comment (NancyLi1013) [inline ports/asio/portfile.cmake:11]: Could you please use `vcpkg_replace_string() `to update these codes?

```
file(READ "${SOURCE_PATH}/asio/include/asio/detail/config.hpp" _contents)
string(REPLACE "defined(ASIO_STANDALONE)" "!defined(VCPKG_DISABLE_ASIO_STANDALONE)" _contents "${_contents}")
file(WRITE "${SOURCE_PATH}/asio/include/asio/detail/config.hpp" "${_contents}")
```

Comment (NancyLi1013) [reply ports/asio/portfile.cmake:11]: Could you please update this

```
file(INSTALL
    ${CMAKE_CURRENT_LIST_DIR}/asio-config.cmake
    DESTINATION ${CURRENT_PACKAGES_DIR}/share/asio/
)
```

as

```
file(INSTALL
    ${CMAKE_CURRENT_LIST_DIR}/asio-config.cmake
    DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
)
```
?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
Hi @mathisloge 
Thanks for posting this issue.
The regressions caused by this changes are like this:

For `restbed`

```
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(56): error C2039: 'get_io_service': is not a member of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\installed\x86-windows\include\asio/ip/tcp.hpp(82): note: see declaration of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(56): error C2672: 'std::make_shared': no matching overloaded function found
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(57): error C2039: 'get_io_service': is not a member of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\installed\x86-windows\include\asio/ip/tcp.hpp(82): note: see declaration of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(57): error C2672: 'std::make_shared': no matching overloaded function found
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(125): error C2039: 'get_io_service': is not a member of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\installed\x86-windows\include\asio/ip/tcp.hpp(82): note: see declaration of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(125): error C2530: 'io_service': references must be initialized
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(127): error C3536: 'io_service': cannot be used before it is initialized
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(448): error C2039: 'get_io_service': is not a member of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\installed\x86-windows\include\asio/ip/tcp.hpp(82): note: see declaration of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(448): error C2530: 'io_service': references must be initialized
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(450): error C3536: 'io_service': cannot be used before it is initialized
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(597): error C2039: 'get_io_service': is not a member of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\installed\x86-windows\include\asio/ip/tcp.hpp(82): note: see declaration of 'asio::basic_stream_socket<asio::ip::tcp,asio::any_io_executor>'
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(597): error C2530: 'io_service': references must be initialized
D:\buildtrees\restbed\src\80d5d74d1e-eb47e978c1.clean\source\corvusoft\restbed\detail\socket_impl.cpp(599): error C3536: 'io_service': cannot be used before it is initialized
```
For `opendnp3`

```
D:\buildtrees\opendnp3\src\2.3.2-0b9dada440.clean\cpp\libs\src\asiopal\SerialChannel.cpp(35): error C2039: 'get_io_context': is not a member of 'asio::io_context::strand'
D:\installed\x64-windows\include\asio/io_context.hpp(239): note: see declaration of 'asio::io_context::strand'
D:\buildtrees\opendnp3\src\2.3.2-0b9dada440.clean\cpp\libs\src\asiopal\SerialChannel.cpp(35): error C2512: 'asio::basic_serial_port<asio::any_io_executor>::basic_serial_port': no appropriate default constructor available
```
Could you please try to fix these regressions?

For `opendnp3`, the latest release seems fix this problem. You just need to update it. Please see https://github.com/dnp3/opendnp3/issues/373
For `restbed`,  the fix has not been merged to master yet. Please see https://github.com/Corvusoft/restbed/pull/444

**Comment by mathisloge** [comment]
Hi @NancyLi1013,
Thanks for the Feedback.

Both Regression should be fixed in its own pull request correct?

Should base these new pull requests on this pull request or base them on master?

**Comment by NancyLi1013** [comment]
For `opendnp3`, just need to update to the latest release, which has included this fix.

For `restbed`, please update to the latest version and make a patch to fix this issue based on the pull request.

**Comment by mathisloge** [comment]
@NancyLi1013 i should have clarified this better. I mean if I have to create to new pull requests in vcpkg for the port updates and base these on the asio pull request or should I update both ports in the asio pull request?

**Comment by NancyLi1013** [comment]
> @NancyLi1013 i should have clarified this better. I mean if I have to create two new pull requests in vcpkg for the port updates and base these on the asio pull request or should I update both ports in the asio pull request?

Update these two ports in asio pull request. Just add the changes in this PR.

**Comment by mathisloge** [comment]
okay, I've messed this up. After the rebase everything showed up as changed... 
Opendp3 is fixed. 

Should i close this PR and make a new clean PR?

**Comment by NancyLi1013** [comment]
If you're not clear about how to revert the changes, it would be better to make a new PR to update these ports.

**Pull request closed without merging**